### DISCUSSION TITLE
org.jboss.resteasy.reactive.common.core.BlockingNotAllowedException - Rest Endpoint with a Request Body

### DISCUSSION THREAD

**Discussion by jaiprasad266** [quarkusio/quarkus#25859] [Community]
Hi Everyone ,

I have a quarkus rest application where there is one endpoint of operation GET.
As part of the rest request we are passing a request body and the endpoint has to read this request body and map it to java pojo.
Getting the below exception when request is sent to the endpoint.
Any inputs will be of help . 

Using :  <version>2.9.0.Final</version>

<dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-resteasy-reactive-jackson</artifactId>
    </dependency>

"stack": "org.jboss.resteasy.reactive.common.core.BlockingNotAllowedException: Attempting a blocking read on io thread\r\n\tat org.jboss.resteasy.reactive.server.vertx.VertxInputStream$VertxBlockingInput.readBlocking(VertxInputStream.java:242)\r\n\tat org.jboss.resteasy.reactive.server.vertx.VertxInputStream.readIntoBuffer(VertxInputStream.java:120)\r\n\tat org.jboss.resteasy.reactive.server.vertx.VertxInputStream.read(VertxInputStream.java:82)\r\n\tat com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.ensureLoaded(ByteSourceJsonBootstrapper.java:539)\r\n\tat com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.detectEncoding(ByteSourceJsonBootstrapper.java:133)\r\n\tat com.fasterxml.jackson.core.json.ByteSourceJsonBootstrapper.constructParser(ByteSourceJsonBootstrapper.java:256)\r\n\tat com.fasterxml.jackson.core.JsonFactory._createParser(JsonFactory.java:1655)\r\n\tat com.fasterxml.jackson.core.JsonFactory.createParser(JsonFactory.java:1083)\r\n\tat com.fasterxml.jackson.databind.ObjectReader.createParser(ObjectReader.java:1049)\r\n\tat com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1427)\r\n\tat io.quarkus.resteasy.reactive.jackson.runtime.serialisers.ServerJacksonMessageBodyReader.doReadFrom(ServerJacksonMessageBodyReader.java:62)\r\n\tat io.quarkus.resteasy.reactive.jackson.runtime.serialisers.ServerJacksonMessageBodyReader.readFrom(ServerJacksonMessageBodyReader.java:53)\r\n\tat org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:108)\r\n\tat org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:68)\r\n\tat org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:25)\r\n\tat org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:141)\r\n\tat org.jboss.resteasy.reactive.server.handlers.RestInitialHandler.beginProcessing(RestInitialHandler.java:49)\r\n\tat org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:17)\r\n\tat org.jboss.resteasy.reactive.server.vertx.ResteasyReactiveVertxHandler.handle(ResteasyReactiveVertxHandler.java:7)\r\n\tat io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)\r\n\tat io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)\r\n\tat io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)\r\n\tat io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:83)\r\n\tat io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)\r\n\tat io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)\r\n\tat io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:126)\r\n\tat io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)\r\n\tat io.vertx.ext.web.handler.impl.StaticHandlerImpl.lambda$sendStatic$1(StaticHandlerImpl.java:274)\r\n\tat io.vertx.core.impl.future.FutureImpl$3.onSuccess(FutureImpl.java:141)\r\n\tat io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54)\r\n\tat io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)\r\n\tat io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)\r\n\tat io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)\r\n\tat io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)\r\n\tat io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)\r\n\tat io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\r\n\tat java.base/java.lang.Thread.run(Thread.java:834)"

**Comment by jaiprasad266** [comment]
Below is the request body and the endpoint draft.

    @Path("/data")
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Multi<DataRecord[]> getData(RestRequestBody restRequestBody) throws IOException {

       /* code */

         return multi;
    }

------------------------------------------Request Body Below --------------------	
	
	{
		"id": "1",
		"params": [
        {
            "paramName": "p1",
            "value": "abcd"
        },
        {
            "paramName": "time1",
            "value": "1652133343112"
        },
        {
            "paramName": "time2",
            "value": "1652133345112"
        }
      ]
	}

**Reply by jaiprasad266** [reply]
I have tried with POST Http operation , it is able to read the request body now. Not sure why GET operation is not supporting this.

**Reply by maxandersen** [reply]
Somehow quarkus detects you are trying to do a blocking operation while running a non locking method. What does your code do? 

If you want it blocking don't return multi or mark it as @blocking.

**Reply by JayGhiya** [reply]
yes @maxandersen looks like a blocking operation is in place which will negate the benefit of event loop threadpool. @jaiprasad266  if you could share the workflow that happens when you receive the request. @maxandersen  the problem is why the same logic is working for post. it should complain for post too if it is blocking logic.

**Reply by jaiprasad266** [reply]
@maxandersen  so we a rest endpoint which accepts http request to fetch data from cosmos db .
At the dao layer there the query preparation and execution and resultset processing to java pojo and finally return the result.
We are using the quarkus cassandra client to interact the db and executing the statement and processing the resultset in reactive style.
Below is the doc reference of datastax to execute reactive queries.
https://docs.datastax.com/en/developer/java-driver/4.4/manual/core/reactive/

Also like @JayGhiya  pointed out we are able to process the request body for the POST operation but not for the GET operation.
